src/Components/Server/test/Circuits/CircuitRegistryTest.cs

```diff
@@ -313,16 +314,108 @@ public async Task DisconnectWhenAConnectIsInProgress()
             Assert.False(registry.DisconnectedCircuits.TryGetValue(circuitHost.CircuitId, out _));
         }
 
+        [Fact]
+        public void CircuitRegistryUsesConfiguredMaxRetainedDisconnectedCircuitsValue()
+        {
+            // Arrange
+            var circuitIdFactory = TestCircuitIdFactory.CreateTestFactory();
+            var maxCircuits = 3;
+            var circuitOptions = new CircuitOptions
+            {
+                MaxRetainedDisconnectedCircuits = maxCircuits,
+            };
+            var registry = new TestCircuitRegistry(circuitIdFactory, circuitOptions);
+            var hosts = Enumerable.Range(0, maxCircuits + 2)
+                .Select(_ => TestCircuitHost.Create())
+                .ToArray();
+
+            // Act
+            for (var i = 0; i < hosts.Length; i++)
+            {
+                registry.RegisterDisconnectedCircuit(hosts[i]);
+            }
+
+            // Assert
+            for (var i = 0; i < maxCircuits; i++)
+            {
+                Assert.True(registry.DisconnectedCircuits.TryGetValue(hosts[i].CircuitId, out var _));
+            }
+
+            // Additional circuits do not get registered.
```

Comment:
We need to rethink this behavior.

Reply:
Yup. Maybe the cache needs a LRU policy. Right now it effectively always rejects newer adds: https://github.com/aspnet/Extensions/blob/master/src/Caching/Memory/src/MemoryCache.cs#L156.

Reply:
@Tratcher \ @JunTaoLuo \ @davidfowl  what do you think about having a eviction strategy for capacity in the cache? All entries being equal priority, we would want the cache to do knock things based on LRU. Should we add a feature to the cache or roll our own thing?

Reply:
LRU is already part of the eviction strategy.

Reply:
https://github.com/aspnet/Extensions/blob/f12d709976e382672ce100bc00381a8847f06489/src/Caching/Memory/src/MemoryCache.cs#L460-L461